### PR TITLE
Adds land-mask option to threshold CLI

### PR DESCRIPTION
**Pull request by MoseleyS**
Addresses #1557 

OccurrenceWithinVicinity has a land-mask-cube option which treats land and sea points separately.

Threshold CLI has an option to apply a land-mask cube when in-vicinity processing is in use.

I have also converted the existing tests into parametrized pytests as this made adding new tests easier, especially for the acceptance tests. This makes the PR look a lot bigger than it really is.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver_tests/acceptance/test_threshold.py [hunk 1]
```diff
@@ -41,114 +41,57 @@
 run_cli = acc.run_cli(CLI)
 
 
-def test_basic(tmp_path):
-    """Test basic thresholding"""
-    kgo_dir = acc.kgo_root() / "threshold/basic"
+@pytest.mark.parametrize(
+    "extra_args,kgo_subdir",
```
Comment (gavinevans) [inline improver_tests/acceptance/test_threshold.py:45]: I'm not sure that we've made such large scale use of the parametrize functionality in the acceptance tests. I'm not sure if there might be readability concerns, although we do do things like this in the unit tests, I think. Do you have any thoughts?

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 2]
```diff
@@ -44,187 +44,236 @@
...
+    mask[0, 4] = 1
+    cube.data = np.ma.array(cube.data, mask=mask)
+    result = OccurrenceWithinVicinity(DISTANCE).maximum_within_vicinity(cube)
+    assert isinstance(result, Cube)
+    assert isinstance(result.data, np.ma.core.MaskedArray)
+    assert np.allclose(result.data.data, expected)
+    assert np.allclose(result.data.mask, mask)
+
+
+def test_with_land_mask(cube, land_mask_cube):
```
Comment (gavinevans) [inline improver_tests/utilities/test_OccurrenceWithinVicinity.py:177]: As a note for another reviewer, I think that only test_with_land_mask, test_with_land_mask_and_mask, test_with_invalid_land_mask_name and test_with_invalid_land_mask_coords have been added, whilst the other tests have only been refactored.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @MoseleyS 👍 

I've added a few minor comments.

File: improver/cli/threshold.py [hunk 3]
```diff
@@ -46,6 +46,7 @@ def process(
     fuzzy_factor: float = None,
     collapse_coord: str = None,
     vicinity: float = None,
+    land_mask: cli.inputcube = None,
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/threshold.py:49]: Of course. I wasn't aware of that, but very happy to be standard!

File: improver/cli/threshold.py [hunk 4]
```diff
@@ -139,7 +143,11 @@ def process(
 
     if vicinity is not None:
         # smooth thresholded occurrences over local vicinity
-        each_threshold_func_list.append(OccurrenceWithinVicinity(vicinity))
+        each_threshold_func_list.append(
+            OccurrenceWithinVicinity(vicinity, land_mask_cube=land_mask)
+        )
+    elif land_mask:
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/cli/threshold.py:149]: The opening `if vicinity is not None` comes from before this PR and I think is meant to differentiate between a vicinity that is unset and a vicinity that is set to `0.0`.
There's no need for the `elif` to have `not vicinity` as this is implicit from the `if` test.

File: improver_tests/acceptance/test_threshold.py [hunk 5]
```diff
@@ -41,114 +41,57 @@
 run_cli = acc.run_cli(CLI)
 
 
-def test_basic(tmp_path):
-    """Test basic thresholding"""
-    kgo_dir = acc.kgo_root() / "threshold/basic"
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/acceptance/test_threshold.py:44]: Yes, and also test_threshold_config (was at the bottom).

File: improver_tests/acceptance/test_threshold.py [hunk 6]
```diff
@@ -197,31 +140,30 @@ def test_collapse_realization_masked_data(tmp_path):
     acc.compare(output_path, kgo_path)
 
 
-def test_vicinity(tmp_path):
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/acceptance/test_threshold.py:143]: Yes

File: improver_tests/acceptance/test_threshold.py (hunk 1 above)
Comment (MoseleyS) [reply improver_tests/acceptance/test_threshold.py:45]: I like it as I am more confident that the things that should be consistent between tests and the things that should change are clearly separated.

**Review by gavinevans — DISMISSED** [review]

**Review by fionaRust — APPROVED** [review]
I'm happy with this change, just put one docstring comment on

**Review by gavinevans — APPROVED** [review]

**Pull request merged**